src/V8JsiRuntime_impl.h

```diff
@@ -640,11 +644,38 @@ class V8Runtime : public facebook::jsi::Runtime {
   static void GCEpilogueCallback(v8::Isolate *isolate, v8::GCType type, v8::GCCallbackFlags flags);
 
  private:
+ // RAII wrappers for multi threaded support - order of the various scopes matters
+  struct IsolateScope {
+    IsolateScope(const V8Runtime *runtime) : 
+      _isolate_scope(runtime->GetIsolate()),
+      _handle_scope(runtime->GetIsolate()),
+      _context_scope(runtime->GetContextLocal())
+    {}
+
+  protected:
+    v8::Isolate::Scope _isolate_scope;
+    v8::HandleScope _handle_scope;
+    v8::Context::Scope _context_scope;
+  };
+
+  struct IsolateLocker {
+    IsolateLocker(const V8Runtime *runtime) {
+      if (runtime->args_.flags.enableMultiThread) {
+        _locker = std::make_unique<v8::Locker>(runtime->GetIsolate());
+      }
+
+      _isolateScope = std::make_unique<IsolateScope>(runtime);
```

Comment:
>make_unique [](http://example.com/codeflow?start=27&length=11)

Why do we to do 2 memory allocation each time we use the IsolateLocker? Is it possible to use field initializers like in the IsolateScope.

Reply:
Could we use the std::optional for the conditional first case?

Reply:
Note that Hermes has uses templated decorators ([https://github.com/facebook/hermes/blob/main/API/jsi/jsi/threadsafe.h](https://github.com/facebook/hermes/blob/main/API/jsi/jsi/threadsafe.h)) to take locks .. something like this might might help in avoiding locking costs for single threaded scenarios .. 

But it looks like not used in react-native right now ..

RN code explicitly takes locks before executing any JS code, so ensuring only one thread is running at the same time: 
[https://github.com/facebook/react-native/blob/main/ReactCommon/runtimeexecutor/ReactCommon/RuntimeExecutor.h](https://github.com/facebook/react-native/blob/main/ReactCommon/runtimeexecutor/ReactCommon/RuntimeExecutor.h)

Do V8 require taking IsolateLocks if the client code guarantees that only one thread is running JS at a time ?

Reply:
> Note that Hermes has uses templated decorators (https://github.com/facebook/hermes/blob/main/API/jsi/jsi/threadsafe.h) to take locks .. something like this might might help in avoiding locking costs for single threaded scenarios ..
> 
> But it looks like not used in react-native right now ..
> 
> RN code explicitly takes locks before executing any JS code, so ensuring only one thread is running at the same time: https://github.com/facebook/react-native/blob/main/ReactCommon/runtimeexecutor/ReactCommon/RuntimeExecutor.h
> 
> Do we require taking IsolateLocks if the client code guarantees that only one thread is running JS at a time ?

> I don't intent to block this PR .. Just a question :)

Reply:
v8::Locker is not copy constructible nor move constructible, so not sure how I'd optionally initialize a std::optional<v8::Locker>. I want to get rid of the dynamic allocs too; any ideas?

Reply:
If the Locker was ever used, it must always be used on that Isolate (so it doesn't matter that the client code "guarantees" it - it could only be a per-Isolate option).